File: tests/python/test_camera.py
from __future__ import annotations

import numpy as np
import pytest

import svo


def test_look_at_generate_rays_shape_and_center_direction() -> None:
    camera = svo.Camera.look_at(
        origin=np.array([0.0, 0.0, 0.0], dtype=np.float32),
        target=np.array([0.0, 0.0, -1.0], dtype=np.float32),
        up=np.array([0.0, 1.0, 0.0], dtype=np.float32),
        width=5,
        height=5,
        vertical_fov_y_degrees=60.0,
    )

    origins, directions = camera.generate_rays()

    assert origins.shape == (5, 5, 3)
    assert directions.shape == (5, 5, 3)
    assert origins.dtype == np.float32
    assert directions.dtype == np.float32
    np.testing.assert_allclose(origins[2, 2], [0.0, 0.0, 0.0], atol=1e-6)
    np.testing.assert_allclose(directions[2, 2], [0.0, 0.0, -1.0], atol=1e-6)
    np.testing.assert_allclose(np.linalg.norm(directions, axis=2), 1.0, atol=1e-6)


def test_from_intrinsics_and_conventions() -> None:
    intrinsics = svo.CameraIntrinsics(width=3, height=3, fx=2.0, fy=2.0, cx=1.5, cy=1.5)
    opengl = svo.Camera.from_intrinsics(
        [0.0, 0.0, 0.0],
        [0.0, 0.0, -1.0],
        [0.0, 1.0, 0.0],
        intrinsics,
        svo.CameraConvention.OpenGL,
    )
    cv = svo.Camera.from_intrinsics(
        [0.0, 0.0, 0.0],
        [0.0, 0.0, -1.0],
        [0.0, 1.0, 0.0],
        intrinsics,
        svo.CameraConvention.ComputerVision,
    )

    _, gl_directions = opengl.generate_rays()
    _, cv_directions = cv.generate_rays()

    np.testing.assert_allclose(gl_directions[1, 1], cv_directions[1, 1], atol=1e-6)
    assert gl_directions[0, 1, 1] == pytest.approx(-cv_directions[0, 1, 1], abs=1e-6)


def test_camera_bad_inputs_raise_clear_errors() -> None:
    with pytest.raises(svo.ValidationError, match="width and height"):
        svo.Camera.look_at([0, 0, 0], [0, 0, -1], [0, 1, 0], 0, 3, 60.0)

    with pytest.raises(svo.ValidationError, match="vertical_fov_y_degrees"):
        svo.Camera.look_at([0, 0, 0], [0, 0, -1], [0, 1, 0], 3, 3, 180.0)

    with pytest.raises(svo.ValidationError, match="forward vector"):
        svo.Camera.look_at([0, 0, 0], [0, 0, 0], [0, 1, 0], 3, 3, 60.0)

    with pytest.raises(svo.ValidationError, match="up vector"):
        svo.Camera.look_at([0, 0, 0], [0, 0, -1], [0, 0, -1], 3, 3, 60.0)

    with pytest.raises(svo.ValidationError, match="fx and fy"):
        bad = svo.CameraIntrinsics(width=3, height=3, fx=0.0, fy=2.0, cx=1.5, cy=1.5)
        svo.Camera.from_intrinsics([0, 0, 0], [0, 0, -1], [0, 1, 0], bad)

    with pytest.raises(ValueError, match="origin must have shape"):
        svo.Camera.look_at([0, 0], [0, 0, -1], [0, 1, 0], 3, 3, 60.0)

File: tests/cpp/test_camera.cpp
#include <svo/Camera.hpp>
#include <svo/Error.hpp>

#include <glm/geometric.hpp>

#include <cmath>
#include <cstdlib>
#include <exception>
#include <iostream>
#include <string>

namespace {

void require(bool condition, const std::string& message) {
  if (!condition) {
    std::cerr << message << '\n';
    std::exit(1);
  }
}

void require_near(float actual, float expected, float tolerance, const std::string& message) {
  if (std::fabs(actual - expected) > tolerance) {
    std::cerr << message << ": expected " << expected << ", got " << actual << '\n';
    std::exit(1);
  }
}

void require_vec_near(
    const glm::vec3& actual,
    const glm::vec3& expected,
    float tolerance,
    const std::string& message) {
  require_near(actual.x, expected.x, tolerance, message + " x");
  require_near(actual.y, expected.y, tolerance, message + " y");
  require_near(actual.z, expected.z, tolerance, message + " z");
}

template <typename Func>
void require_validation_error(Func&& func, const std::string& message) {
  try {
    func();
  } catch (const svo::ValidationError&) {
    return;
  } catch (const std::exception& error) {
    std::cerr << message << ": expected ValidationError, got " << error.what() << '\n';
    std::exit(1);
  }
  std::cerr << message << ": expected ValidationError\n";
  std::exit(1);
}

std::size_t ray_index(int width, int x, int y) {
  return static_cast<std::size_t>(y) * static_cast<std::size_t>(width) + static_cast<std::size_t>(x);
}

void test_center_pixel_points_at_target() {
  const svo::Camera camera = svo::Camera::look_at(
      {0.0f, 0.0f, 0.0f},
      {0.0f, 0.0f, -1.0f},
      {0.0f, 1.0f, 0.0f},
      5,
      5,
      60.0f,
      svo::CameraConvention::OpenGL);
  const svo::RayBatch rays = svo::generate_rays_cpu(camera);

  require(rays.width == 5 && rays.height == 5, "ray batch should preserve dimensions");
  require(rays.origins.size() == 25 && rays.directions.size() == 25, "ray batch should be flattened");
  require_vec_near(rays.origins[ray_index(5, 2, 2)], {0.0f, 0.0f, 0.0f}, 1.0e-6f, "origin mismatch");
  require_vec_near(
      rays.directions[ray_index(5, 2, 2)],
      {0.0f, 0.0f, -1.0f},
      1.0e-6f,
      "center ray should point at target");
}

void test_basis_is_orthonormal() {
  const svo::Camera camera = svo::Camera::look_at(
      {1.0f, 2.0f, 3.0f},
      {3.0f, 2.5f, 1.0f},
      {0.0f, 1.0f, 0.0f},
      7,
      5,
      55.0f);

  require_near(glm::length(camera.right()), 1.0f, 1.0e-5f, "right length");
  require_near(glm::length(camera.up()), 1.0f, 1.0e-5f, "up length");
  require_near(glm::length(camera.forward()), 1.0f, 1.0e-5f, "forward length");
  require_near(glm::dot(camera.right(), camera.up()), 0.0f, 1.0e-5f, "right/up dot");
  require_near(glm::dot(camera.right(), camera.forward()), 0.0f, 1.0e-5f, "right/forward dot");
  require_near(glm::dot(camera.up(), camera.forward()), 0.0f, 1.0e-5f, "up/forward dot");
}

void test_fov_changes_corner_spread() {
  const svo::Camera narrow = svo::Camera::look_at(
      {0.0f, 0.0f, 0.0f},
      {0.0f, 0.0f, -1.0f},
      {0.0f, 1.0f, 0.0f},
      5,
      5,
      30.0f);
  const svo::Camera wide = svo::Camera::look_at(
      {0.0f, 0.0f, 0.0f},
      {0.0f, 0.0f, -1.0f},
      {0.0f, 1.0f, 0.0f},
      5,
      5,
      90.0f);

  const glm::vec3 narrow_corner = svo::generate_rays_cpu(narrow).directions.front();
  const glm::vec3 wide_corner = svo::generate_rays_cpu(wide).directions.front();
  require(
      std::fabs(wide_corner.x) > std::fabs(narrow_corner.x),
      "wider FOV should increase horizontal corner spread");
  require(
      std::fabs(wide_corner.y) > std::fabs(narrow_corner.y),
      "wider FOV should increase vertical corner spread");
}

void test_conventions_share_center_and_flip_vertical() {
  const svo::Camera opengl = svo::Camera::look_at(
      {0.0f, 0.0f, 0.0f},
      {0.0f, 0.0f, -1.0f},
      {0.0f, 1.0f, 0.0f},
      3,
      3,
      60.0f,
      svo::CameraConvention::OpenGL);
  const svo::Camera cv = svo::Camera::look_at(
      {0.0f, 0.0f, 0.0f},
      {0.0f, 0.0f, -1.0f},
      {0.0f, 1.0f, 0.0f},
      3,
      3,
      60.0f,
      svo::CameraConvention::ComputerVision);

  const svo::RayBatch gl_rays = svo::generate_rays_cpu(opengl);
  const svo::RayBatch cv_rays = svo::generate_rays_cpu(cv);
  require_vec_near(
      gl_rays.directions[ray_index(3, 1, 1)],
      cv_rays.directions[ray_index(3, 1, 1)],
      1.0e-6f,
      "center conventions should match");
  require_near(
      gl_rays.directions[ray_index(3, 1, 0)].y,
      -cv_rays.directions[ray_index(3, 1, 0)].y,
      1.0e-6f,
      "vertical convention should flip");
}

void test_from_intrinsics() {
  const svo::CameraIntrinsics intrinsics{3, 3, 2.0f, 2.0f, 1.5f, 1.5f};
  const svo::Camera camera = svo::Camera::from_intrinsics(
      {1.0f, 2.0f, 3.0f},
      {1.0f, 2.0f, 2.0f},
      {0.0f, 1.0f, 0.0f},
      intrinsics,
      svo::CameraConvention::OpenGL);
  const svo::RayBatch rays = svo::generate_rays_cpu(camera);
  require_vec_near(
      rays.directions[ray_index(3, 1, 1)],
      {0.0f, 0.0f, -1.0f},
      1.0e-6f,
      "intrinsics center ray");
}

void test_invalid_inputs() {
  require_validation_error(
      [] {
        (void)svo::Camera::look_at({0, 0, 0}, {0, 0, -1}, {0, 1, 0}, 0, 3, 60.0f);
      },
      "zero width");
  require_validation_error(
      [] {
        (void)svo::Camera::look_at({0, 0, 0}, {0, 0, -1}, {0, 1, 0}, 3, -1, 60.0f);
      },
      "negative height");
  require_validation_error(
      [] {
        (void)svo::Camera::look_at({0, 0, 0}, {0, 0, -1}, {0, 1, 0}, 3, 3, 0.0f);
      },
      "invalid fov");
  require_validation_error(
      [] {
        (void)svo::Camera::look_at({0, 0, 0}, {0, 0, 0}, {0, 1, 0}, 3, 3, 60.0f);
      },
      "zero forward");
  require_validation_error(
      [] {
        (void)svo::Camera::look_at({0, 0, 0}, {0, 0, -1}, {0, 0, 0}, 3, 3, 60.0f);
      },
      "zero up");
  require_validation_error(
      [] {
        (void)svo::Camera::look_at({0, 0, 0}, {0, 0, -1}, {0, 0, -1}, 3, 3, 60.0f);
      },
      "degenerate up");
  require_validation_error(
      [] {
        const svo::CameraIntrinsics intrinsics{3, 3, 0.0f, 2.0f, 1.5f, 1.5f};
        (void)svo::Camera::from_intrinsics({0, 0, 0}, {0, 0, -1}, {0, 1, 0}, intrinsics);
      },
      "non-positive fx");
}

}  // namespace

int main() {
  test_center_pixel_points_at_target();
  test_basis_is_orthonormal();
  test_fov_changes_corner_spread();
  test_conventions_share_center_and_flip_vertical();
  test_from_intrinsics();
  test_invalid_inputs();
  return 0;
}
